### ISSUE TITLE
[Bug]: Bearer auth doesn't work with docker and skopeo login

### ISSUE THREAD

**Issue by sagikazarmark** [project-zot/zot#1125]
### zot version

v1.4.3

### Describe the bug

When trying to use the bearer authn feature with an external authorization server and docker/skopeo login, I noticed that login always fails.

First, I suspected an incompatibility between Zot and the custom authorization server (https://github.com/distribution-auth/auth), but the authz server actually returned a valid token.

Then I started some manual testing checking out the code I think I figured out an incompatibility between the bearer auth implementation and docker/skopeo login.

Both docker/skopeo login go to the registry without a token first. That results in the following HTTP response from Zot:

```
❯ curl -vvv http://127.0.0.1:5001/v2/
*   Trying 127.0.0.1:5001...
* Connected to 127.0.0.1 (127.0.0.1) port 5001 (#0)
> GET /v2/ HTTP/1.1
> Host: 127.0.0.1:5001
> User-Agent: curl/7.85.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< Access-Control-Allow-Headers: Authorization,content-type
< Access-Control-Allow-Methods: HEAD,GET,POST,OPTIONS
< Access-Control-Allow-Origin: *
< Content-Type: application/json
< Www-Authenticate: Bearer realm="http://localhost:8080/token",service="localhost:5000",scope="repository::pull"
< Date: Thu, 19 Jan 2023 19:19:42 GMT
< Content-Length: 283
< 
* Connection #0 to host 127.0.0.1 left intact
{"errors":[{"code":"UNAUTHORIZED","message":"authentication required","description":"The access controller was unable to authenticate the client.\n\t\t\tOften this will be accompanied by a Www-Authenticate HTTP response header\n\t\t\tindicating how to authenticate.","detail":null}]}
```

Note the `Www-Authenticate` header.

Docker/skopeo are supposed to go to the authorization service (realm) and send an authorization request. The authorization service running at `localhost:8080` receives the request, but it looks like the scope parameter is not sent to the service upon login. When comparing Zot to the official Docker registry that's in fact the primary difference between the response of the two:

```
❯ curl -vvv localhost:5000/v2/
*   Trying 127.0.0.1:5000...
* Connected to localhost (127.0.0.1) port 5000 (#0)
> GET /v2/ HTTP/1.1
> Host: localhost:5000
> User-Agent: curl/7.85.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< Content-Type: application/json; charset=utf-8
< Docker-Distribution-Api-Version: registry/2.0
< Www-Authenticate: Bearer realm="http://host.docker.internal:8080/token",service="localhost:5000"
< X-Content-Type-Options: nosniff
< Date: Thu, 19 Jan 2023 19:23:38 GMT
< Content-Length: 87
< 
{"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":null}]}
* Connection #0 to host localhost left intact
```

As you can see, there is no scope parameter returned by the official Docker registry.

My theory is that upon login, none of the clients expect to receive a scope and as a result it's not sent to the authorization server.

The debug output of `skopeo login` seems to support that theory:

```
❯ skopeo --debug login -u user -p password --tls-verify=false -v 127.0.0.1:5001
DEBU[0000] Loading registries configuration "/etc/containers/registries.conf" 
DEBU[0000] No credentials matching 127.0.0.1:5001 found in /Users/marksk/.config/containers/auth.json 
DEBU[0000] No credentials matching 127.0.0.1:5001 found in /Users/marksk/.config/containers/auth.json 
DEBU[0000] No credentials matching 127.0.0.1:5001 found in /Users/marksk/.docker/config.json 
DEBU[0000] No credentials matching 127.0.0.1:5001 found in /Users/marksk/.dockercfg 
DEBU[0000] No credentials for 127.0.0.1:5001 found      
DEBU[0000] Looking for TLS certificates and private keys in /etc/docker/certs.d/127.0.0.1:5001 
DEBU[0000] GET https://127.0.0.1:5001/v2/               
DEBU[0000] Ping https://127.0.0.1:5001/v2/ err Get "https://127.0.0.1:5001/v2/": http: server gave HTTP response to HTTPS client (&url.Error{Op:"Get", URL:"https://127.0.0.1:5001/v2/", Err:(*errors.errorString)(0x1400025c030)}) 
DEBU[0000] GET http://127.0.0.1:5001/v2/                
DEBU[0000] Ping http://127.0.0.1:5001/v2/ status 401    
DEBU[0000] GET http://localhost:8080/token?account=user&service=localhost%3A5000 
DEBU[0000] GET http://127.0.0.1:5001/v2/                
DEBU[0000] error logging into "127.0.0.1:5001": unable to retrieve auth token: invalid username/password: unauthorized 
FATA[0000] logging into "127.0.0.1:5001": invalid username/password 
```

Since there is no scope parameter sent to the authorization server by the client, the returned access token won't have any access entries in it either.

That results in authentication always failing upon docker/skopeo login [here](https://github.com/project-zot/zot/blob/v1.4.3/pkg/api/authn.go#L60).

### To reproduce

```json
{
  "distSpecVersion": "1.0.1-dev",
  "storage": {
    "rootDirectory": "/tmp/zot"
  },
  "http": {
    "address": "0.0.0.0",
    "port": "5001",
    "auth": {
      "bearer": {
        "realm": "http://localhost:8080/token",
        "service": "localhost:5000",
        "cert": "distribution-auth/auth/etc/zot/root.crt"
      }
    }
  },
  "log": {
    "level": "debug"
  }
}
```

Tools: docker, skopeo

### Expected behavior

Login should succeed

### Additional context

There is actually a test for the bearer auth, but it does not emulate the client behavior properly: the test assumes that the scope will be properly sent to the authorization header upon login as well (which does not seem to be the case).

A potential solution is handling requests sent to `/v2/` (ie. no specific resource is being accessed) separately.

**Comment by sagikazarmark** [comment]
The bearer auth implementation uses chartmuseum/auth that claims being based on the distribution auth spec: https://github.com/distribution/distribution/blob/main/docs/spec/auth/token.md

That spec explicitly says that the scope may be empty.

**Comment by rchincha** [comment]
@sagikazarmark can you summarize what needs to be fixed in chartmuseum so we can post a PR upstream?

**Comment by sagikazarmark** [comment]
I came to the conclusion that the bearer auth feature is currently broken and it probably doesn't work with any of the non-library clients (it certainly doesn't work with docker and skopeo).

Please see my analysis in the linked PR.

Happy to work with you on a fix, although I'm not sure it's going to be as simple as fixing the chartmusuem upstream. I'd also say it's probably a risk at this point to continue relying on that library.

**Comment by sagikazarmark** [comment]
Friendly ping @rchincha

I've provided detailed information in the above PR about the potential solutions. In short:

- Drop using the chartmuseum library in favor of an alternative
- Fork the chartmuseum library and fix it

I doubt the chartmuseum auth library is maintained at this point. Even if it is, the chartmuseum auth standard is **different** from the docker one.

I'm aware of the new WG Auth, but it's gonna take years until any solution proposed by it can become a standard.

**Comment by anoldguy** [comment]
Oof. Just found this issue. That explains my problems attempting to use [docker_auth](https://github.com/cesanta/docker_auth/) as a token backend for this. Bleh. Thank you, @sagikazarmark for troubleshooting this, hope there's some movement on it. In the meantime, we'll have to stick with basic auth.

**Comment by sagikazarmark** [comment]
@anoldguy I have a patched version somewhere that I used for a demo, let me look it up.

I agree, the token backend is completely broken in Zot at the moment which is a shame.

I don't think fixing the problem is that complicated, but without a decision from the maintainers on the direction any attempts to move in one or the other is a potential waste of time.

Shameless self-promotion: you might want to give Portward a try: https://github.com/portward

It's a token authorization server built from scratch with extensibility in mind. I've also spent a considerable amount of time to gracefully handle the different implementations (the "spec" is unfortunately not very well defined in some edge cases).

Would love to hear your thoughts!

**Comment by sagikazarmark** [comment]
@anoldguy Portward quickstart and demo: https://github.com/portward/quickstart (it contains a patched version of Zot with a quick fix for the token auth method)

**Comment by andaaron** [comment]
An updated version of #1814 was merged in #2176

**Comment by sagikazarmark** [comment]
Awesome!

I did a quick test and it looks like it's indeed working now.
